enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/state/snapshot/PersistentSnapshotDownloader.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright (c) 2002-2017 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.causalclustering.core.state.snapshot;
+
+import java.util.concurrent.TimeUnit;
+
+import org.neo4j.causalclustering.catchup.storecopy.StoreCopyFailedException;
+import org.neo4j.causalclustering.core.consensus.LeaderLocator;
+import org.neo4j.causalclustering.core.consensus.NoLeaderFoundException;
+import org.neo4j.causalclustering.core.state.CommandApplicationProcess;
+import org.neo4j.causalclustering.helper.ExponentialBackoffStrategy;
+import org.neo4j.causalclustering.helper.TimeoutStrategy;
+import org.neo4j.logging.Log;
+
+class PersistentSnapshotDownloader implements Runnable
+{
+    static final String OPERATION_NAME = "download of snapshot";
+
+    private final CommandApplicationProcess applicationProcess;
+    private final LeaderLocator leaderLocator;
+    private final CoreStateDownloader downloader;
+    private final Log log;
+    private final TimeoutStrategy.Timeout timeout;
+    private State state;
+
+    PersistentSnapshotDownloader( LeaderLocator leaderLocator,
+            CommandApplicationProcess applicationProcess, CoreStateDownloader downloader, Log log,
+            TimeoutStrategy.Timeout timeout )
+    {
+        this.applicationProcess = applicationProcess;
+        this.leaderLocator = leaderLocator;
+        this.downloader = downloader;
+        this.log = log;
+        this.timeout = timeout;
+        this.state = State.INITIATED;
+    }
+
+    PersistentSnapshotDownloader( LeaderLocator leaderLocator,
+            CommandApplicationProcess applicationProcess, CoreStateDownloader downloader, Log log )
+    {
+        this( leaderLocator, applicationProcess, downloader, log,
+                new ExponentialBackoffStrategy( 1, 30, TimeUnit.SECONDS ).newTimeout() );
+    }
+
+    private enum State
+    {
+        INITIATED,
+        RUNNING,
+        STOPPED,
+        COMPLETED
+    }
+
+    @Override
+    public void run()
+    {
+        if ( !initialStateOk() )
+        {
+            return;
+        }
+        try
+        {
+            state = State.RUNNING;
```

Comment:
This variable state is accessed from two threads so you need some type of synchronization mechanism. Might be enough to make it volatile. I think it can be good to split it up into two variables, state and keepRunning, because right now you have two threads updating the same variable and their interactions are hard to follow. The variable state can be updated by PersistentSnaptshotDownloader#run and keepRunning can be updated from #stop. The while-loop then checks keepRunning.